spec/controllers/account_activations_controller_spec.rb

```diff
@@ -0,0 +1,52 @@
+describe AccountActivationsController do
+  describe '#edit' do
+    context 'with inactive user' do
+      let(:artist) { FactoryGirl.create(:artist) }
```

Comment:
## FactoryGirl

FactoryGirl is a gem for creating 'factories' to create models. Their [project page](https://github.com/thoughtbot/factory_girl) explains a bunch. I'm also using [faker](https://github.com/stympy/faker) to generate placeholder names, etc.

If you look at `spec/factories/*.rb` you can see the factories I wrote for this project. A nice feature is traits which let you mix and match things you want a model to be. Example:

```ruby
FactoryGirl.create(:artist)
#=> Creates a default artist with faked data

FactoryGirl.create(:artist, :activated)
#=> creates an 'activated' artist, you can see the details of the 'activated' trait
# in `spec/factories/artist.rb`, look for the `trait :activated` block
```

If you need to create a user with some specific traits you can easily override certain values.

```ruby
FactoryGirl.create(:artist, :activated, email: 'test@example.com')
```

FactoryGirl can also create the associations for a model, for example because a `Grant` requires an `Artist` and I have a factory for both and have the line `artist` in the `Grant` factory this will happen:

```ruby
grant = FactoryGirl.create(:grant)
grant.artist # => is an Artist using the default Artist factory
```

Using factories instead of fixtures makes your tests have less global state. For example with your previous tests changing the fixtures by adding another `Voter` would have broken the test of `AdminsController#assign`. Now I specifically create the required models for only that test. This is especially important on projects with many developers. Having fragile tests slows everyone down.

## Rspec

Rspec is a "behavior-driven test framework". There are a few of them for different languages (eg. [Jasmine](https://jasmine.github.io/2.5/introduction] for JavaScript).

For example before one of your tests looked like this:

```ruby
test "artist_discuss" do
  @artist = artists(:michael)
  sign_in_artist(@artist, 'password')

  # bad id
  get grant_submissions_discuss_path(id: 20, authenticity_token: "abc")
  assert_redirected_to "/"

  # not ours
  get grant_submissions_discuss_path(id: 2, authenticity_token: "abc")
  assert_redirected_to "/"

  # good id
  get grant_submissions_discuss_path(id: 1, authenticity_token: "abc")
  assert_template 'grant_submissions/discuss'
end
```

There are a couple of things about this code that aren't great:
* tests of different things aren't separated, if it fails on the "bad id" part but the "good id" part still works the test summary won't tell me. If any of these tests started to fail I would have to read the test code to figure out where to look in the actual code
* The automatic test summary isn't very useful, your comments about what is being tested are comments in the code instead of useable strings to be printed out

After being re-written in rspec we get better summaries and it is more clear what functionality has failed when there is a test failure.

```ruby
describe '#discuss' do
  context 'artist signed in' do
    let!(:artist) { FactoryGirl.create(:artist) }

    before { sign_in_artist(artist.id) }

    context 'with non-existent grant_submission id' do
      it 'redirects to /' do
        get 'discuss', { id: GrantSubmission.count + 1 }
        expect(response).to redirect_to('/')
      end
    end

    context 'with grant_submission id for another artist' do
      let!(:another_grant_submission) { FactoryGirl.create(:grant_submission) }

      it 'redirects to /' do
        get 'discuss', { id: another_grant_submission.id }
        expect(response).to redirect_to('/')
      end
    end

    context 'with valid submission for artist' do
      it 'shows discussion page' do
        get 'discuss', { id: grant_submission.id }
        expect(response).to render_template('grant_submissions/discuss')
      end
    end
  end
end
```

The above examples creates this summary (with color highlighting if configured!)

```
GrantSubmissionsController
  #discuss
    artist signed in
      with non-existent grant_submission id
        redirects to /
      with grant_submission id for another artist
        redirects to /
      with valid submission for artist
        shows discussion page
```

I try and follow the guidelines at [betterspecs.org](http://betterspecs.org/).

I'm not using any fancy matchers here but there are things that can save you some lines. For example this kind of code

```ruby
assert_equal Grant.count, 0
post create_grant(artist_id: 1)
assert_equal Grant.count 2
```

Can be changed to ~look like this and will give a better message if Grant.count doesn't change by one (also this test would no longer require the count to from from 0 to 1, it could go from 4 to 5):

```ruby
expect { post create_grant(artist_id: 1) }.to change { Grant.count }.by(1)
```

Definitely this description isn't all you would need to start writing rspec, for example there is a distinction between `let` and `let!` that is often important in controller tests. Also there can be some gotchas like `expect { do_somthing }.to change { artist.email }` (will likely fail) vs `expect { do_somthing }.to change { artist.reload.email }` (probably what you actually want).